### PR TITLE
Adding DSS capabilities

### LINKED ISSUE #1617: [TASK] Adding DSS Capabilities to Validation Code
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Adding Dynamical System Scaling to Postprocessor for Validation

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by yoshiurr-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1617

##### What are the significant changes in functionality due to this change request?
- The main function of this pull request is to add Dynamical System Scaling (DSS) to wangc/validation_pp_base.
- The addition of metric and postprocessor DSS code is part of the PR.
- To pass DSS specific data to metric code DSS, the MetricDistributor.py has been modified. When "DSS" is added as the metric in the xml file, the modified code will pass postprossed data to only the DSS metric code.
- Test files showing the DSS functionality is added as well.
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Models/PostProcessors/validationAlgorithms/PPDSS.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+    """
+      This method executes the postprocessor action. In this case it loads the
+      results to specified dataObject
+      @ In, inputIn, list, dictionary of data to process
+      @ Out, outputDict, dict, dictionary containing the post-processed results
+    """
+    # assert
+    assert(isinstance(inputIn, list))
+    assert(isinstance(inputIn[0], xr.Dataset) or isinstance(inputIn[0], DataObjects.DataSet))
+    # the input can be either be a list of dataobjects or a list of datasets (xarray)
```
Comment (Jimmy-INL) [inline framework/Models/PostProcessors/validationAlgorithms/PPDSS.py:166]: These asserts will error out!
If they are necessary, maybe `assert(isinstance(inputIn['Data][0][-1],xr.Dataset))` ?

File: framework/Models/PostProcessors/validationAlgorithms/PPDSS.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+  def run(self, inputIn):
+    """
+      This method executes the postprocessor action. In this case it loads the
+      results to specified dataObject
+      @ In, inputIn, list, dictionary of data to process
+      @ Out, outputDict, dict, dictionary containing the post-processed results
+    """
+    # assert
+    assert(isinstance(inputIn["Data"], list))
+    assert(isinstance(inputIn["Data"][0][-1], xr.Dataset) and isinstance(inputI["Data"][1][-1], xr.Dataset))
```
Comment (Jimmy-INL) [inline framework/Models/PostProcessors/validationAlgorithms/PPDSS.py:165]: There is a missing `n` in `inpoutI['Data'][1][-1]`

Comment (Jimmy-INL) [reply framework/Models/PostProcessors/validationAlgorithms/PPDSS.py:165]: I tried to push it my self but it gave me:
`remote: Password authentication is temporarily disabled as part of a brownout. Please use a personal access token instead.
remote: Please see https://github.blog/2020-07-30-token-authentication-requirements-for-api-and-git-operations/ for more information.
fatal: unable to access 'https://github.com/idaholab/raven.git/': The requested URL returned error: 403`

**Comment by wangcj05** [comment]
@yoshiurr-INL Can you rebase your branch with devel and try to resolve the conflicts?

**Comment by wangcj05** [comment]
@yoshiurr-INL @Jimmy-INL This branch has been cleaned up.

**Comment by yoshiurr-INL** [comment]
> @yoshiurr-INL Is it ready to be reviewed?

I will finish the documentation today. Afterwards, it should be ready for review.

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+from utils import utils
+from utils import InputData, InputTypes
+from utils import xmlUtils
+import Files
+import DataObjects
+from ..ValidationBase import ValidationBase
+#import Distributions
+#import MetricDistributor
+#from .ValidationBase import ValidationBase
+#from Models.PostProcessors import PostProcessor
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:42]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for
+        specifying input of cls.
+    """
+    inputSpecification = super(PPDSS, cls).getInputSpecification()
+    #featuresInput = InputData.parameterInputFactory("Features", contentType=InputTypes.StringListType)
+    #featuresInput.addParam("type", InputTypes.StringType)
+    #inputSpecification.addSub(featuresInput)
+    #targetsInput = InputData.parameterInputFactory("Targets", contentType=InputTypes.StringListType)
+    #targetsInput.addParam("type", InputTypes.StringType)
+    #inputSpecification.addSub(targetsInput)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:64]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+    multiOutputInput = InputData.parameterInputFactory("multiOutput", contentType=InputTypes.StringType)
+    inputSpecification.addSub(multiOutputInput)
+    multiOutput = InputTypes.makeEnumType('MultiOutput', 'MultiOutputType', 'raw_values')
+    multiOutputInput = InputData.parameterInputFactory("multiOutput", contentType=multiOutput)
+    inputSpecification.addSub(multiOutputInput)
+    #pivotParameterInput = InputData.parameterInputFactory("pivotParameter", contentType=InputTypes.StringType)
+    #inputSpecification.addSub(pivotParameterInput)
+    #
+    # Have added the new pivotParameters for feature and target. The original has been commented out.
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:73]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+    pivotParameterFeatureInput = InputData.parameterInputFactory("pivotParameterFeature", contentType=InputTypes.StringType)
+    inputSpecification.addSub(pivotParameterFeatureInput)
+    pivotParameterTargetInput = InputData.parameterInputFactory("pivotParameterTarget", contentType=InputTypes.StringType)
+    inputSpecification.addSub(pivotParameterTargetInput)
+    #metricInput = InputData.parameterInputFactory("Metric", contentType=InputTypes.StringType)
+    #metricInput.addParam("class", InputTypes.StringType, True)
+    #metricInput.addParam("type", InputTypes.StringType, True)
+    #inputSpecification.addSub(metricInput)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:81]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+    scaleTypeInput = InputData.parameterInputFactory("scale", contentType=InputTypes.StringType)
+    inputSpecification.addSub(scaleTypeInput)
+    scaleRatioBetaInput = InputData.parameterInputFactory("scaleBeta", contentType=InputTypes.FloatListType)
+    inputSpecification.addSub(scaleRatioBetaInput)
+    scaleRatioOmegaInput = InputData.parameterInputFactory("scaleOmega", contentType=InputTypes.FloatListType)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:86]: please use "descr" key words to add the descriptions for each node, and attribute in the XML.

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+      @ Out, None
+    """
+    super().__init__()
+    self.printTag = 'POSTPROCESSOR DSS Scaling and Metrics'
+    self.name = 'PPDSS'
+    self.dynamic               = True  # is it time-dependent?
+    self.dynamicType = ['dynamic']
+    self.features              = None  # list of feature variables
+    self.targets               = None  # list of target variables
+    #self.metricsDict           = {}    # dictionary of metrics that are going to be assembled
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:103]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+    self.multiOutput           = 'raw_values' # defines aggregating of multiple outputs for HistorySet
+                                # currently allow raw_values
+    #self.pivotParameter        = None  # list of pivot parameters
+    #self.pivotValues           = []
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:107]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+    self.pivotParameterFeature = None
+    self.pivotValuesFeature    = []
+    self.pivotParameterTarget  = None
+    self.pivotValuesTarget     = []
+    self.scaleType             = None
+    #self.processTimeRecord     = []
+    #self.scaleType             = ['DataSynthesis','2_2_Affine','Dilation','beta_strain','omega_strain','identity']
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:114]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+    # assembler objects to be requested
+    self.scaleRatioBeta        = []
+    self.scaleRatioOmega       = []
+    #self.addAssemblerObject('Metric', InputData.Quantity.one_to_infinity)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:118]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+      @ Out, None
+    """
+    super()._handleInput(paramInput)
+    for child in paramInput.subparts:
+      if child.getName() == 'Metric':
+        if 'type' not in child.parameterValues.keys() or 'class' not in child.parameterValues.keys():
+          self.raiseAnError(IOError, 'Tag Metric must have attributes "class" and "type"')
+      elif child.getName() == 'Features':
+        self.features = child.value
+        #self.featuresType = child.parameterValues['type']
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:133]: remove?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+      elif child.getName() == 'scaleBeta':
+        self.scaleRatioBeta = child.value
+      elif child.getName() == 'scaleOmega':
+        self.scaleRatioOmega = child.value
+      else:
+        self.raiseAnError(IOError, "Unknown xml node ", child.getName(), " is provided for metric system")
+    #if not self.features:
+    #  self.raiseAnError(IOError, "XML node 'Features' is required but not provided")
+    #elif len(self.features) != len(self.targets):
+    #  self.raiseAnError(IOError, 'The number of variables found in XML node "Features" is not equal the number of variables found in XML node "Targets"')
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:154]: remove the commented codes

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+    datasets = [data for _, _, data in inputIn['Data']]
+    #print("datasets:",datasets)
+    names = []
+    pivotParameterTarget = self.pivotParameterTarget
+    pivotParameterFeature = self.pivotParameterFeature
+    names = [self.getDataSetName(inp[-1]) for inp in inputIn['Data']]
+    #print("names:",names)
+    #print("inputIn:",inputIn)
+    #print("inputIn['Data'][0][2].indexes:",inputIn['Data'][0][2].indexes)
+    #print("inputIn['Data'][0][-1].indexes:",inputIn['Data'][0][-1].indexes)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:177]: remove the commented lines of codes

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+    return evaluation
+
+  def _evaluate(self, datasets, **kwargs):
+    """
+      Main method to "do what you do".
+      @ In, datasets, list, list of datasets (data1,data2,etc.) to used.
+      @ In, kwargs, dict, keyword arguments
+      @ Out, outputDict, dict, dictionary containing the results {"feat"_"target"_"metric_name":value}
+    """
+    #print("datasets:",datasets)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:206]: remove this line

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+        pivotSize = pivotTargetSize
+      else:
+        pivotSize = pivotFeatureSize
+
+      if pivotFeatureSize == pivotSize:
+        y_count = featData.shape[0]
+        z_count = featData.shape[1]
+      else:
+        y_count = targData.shape[0]
+        z_count = targData.shape[1]
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:259]: camelBack for y_count and z_count

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+      for cnt2 in range(y_count):
+        if pivotFeatureSize == pivotSize:
+          featureBeta[cnt2] = featData[cnt2]
+          interpGrid = pivotFeature
+        else:
+          interpFunction = interp1d(pivotFeature,featData[cnt2],kind='linear',fill_value='extrapolate')
+          interpGrid = timeScalingRatio*pivotTarget
+          featureBeta[cnt2] = interpFunction(interpGrid)
+        featureOmega = np.gradient(featureBeta[cnt2],interpGrid)
+        #print("featureOmega:",featureOmega)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:276]: remove

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+        # Excluding NaN type data and exclude corresponding time in grid ina
+        # preperation for numpy simpson integration
+        count=0
+        for i in range(len(featureD[cnt2])):
+          if np.isnan(featureD[cnt2][i])==False and np.isinf(featureD[cnt2][i])==False:
+            count += 1
+        if count > 0:
+          featureInt_new = np.zeros(count)
+          interpGrid_new = np.zeros(count)
+          track_count = 0
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:295]: camelBack for the variables with underscore

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+          for i in range(len(featureD[cnt2])):
+            if np.isnan(featureD[cnt2][i])==False and np.isinf(featureD[cnt2][i])==False:
+              interpGrid_new[track_count] = interpGrid[i]
+              featureInt_new[track_count] = featureInt[i]
+              track_count += 1
+            else:
+              featureD[cnt2][i] = 0
+        #
+        #print("featureInt_new:",featureInt_new)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:304]: Please remove all the print statement in the entire code

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+        # Excluding NaN type data and exclude corresponding time in grid in
+        # preperation for numpy simpson integration
+        count=0
+        for i in range(len(targetD[cnt2])):
+          if np.isnan(targetD[cnt2][i])==False and np.isinf(targetD[cnt2][i])==False:
+            count += 1
+        if count > 0:
+          targetInt_new = np.zeros(count)
+          interpGrid_new = np.zeros(count)
+          track_count = 0
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:343]: please change the variables using underscore to camelBack for the entire file

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+    #---------------
+    for cnt in range(len(realization_array[0])):
+      out = {}
+      for cnt2 in range(len(realization_array)):
+        for key, val in realization_array[cnt2][cnt].items():
+          out[key] = val
+      realizations.append(out)
+    return realizations
+
+  def _getDataFromDatasets(self, datasets, var, names=None):
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:411]: Is it function used by other post-processors, like "Probabilistic"? If so, please move it to the Validation base class

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@yoshiurr-INL I have several comments for your consideration. In addition:
1. Please clean up the csv files in both DSS and gold/DSS folders 
2. Please update "tests" file with DSS, otherwise your test will not be tested by the machine
3. Please update the manual with some detailed information or add a user guide in another PR
4. please remove print statement, commented lines, and use camelback instead of underscores for the variables.

File: doc/user_manual/metrics.tex [hunk 22]
```diff
@@ -599,3 +598,32 @@ \subsubsection{Distance Based Metric}
...
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsection{Dynamical System Scaling}
+\label{subsection:DSS}
+The Dynamical System Scaling (DSS) is a distance metrics that is used to measure the separation
+between two time-dependent data sets.
+
+The specifications of a DSS metric is defined within the \xmlNode{Metric} XML block. The attribute \xmlAttr{subType} must be \textbf{PPDSS} (see \ref{subsubsec:Validation}) in the \xmlNode{PostProcessor} for the outputs of the post-processor to be in the right format for DSS metric inputs.
+
+This XML node needs to contain the attributes:
```
[the remark this answers is not in the document]
Comment (yoshiurr-INL) [reply doc/user_manual/metrics.tex:610]: Fixed

File: doc/user_manual/postprocessor.tex [hunk 23]
```diff
@@ -1649,13 +1649,36 @@ \subsubsection{Validation}
...
+      \item \xmlNode{multiOutput}, \xmlDesc{string, required field}, to extract raw values for the HistorySet. The user must use ‘raw values’ for the full set of metrics’ calculations to be dumped.
+      \item \xmlNode{scale}, \xmlDesc{string, required field}, specifies the type of time scaling. The following are the options for scaling:
+        \begin{itemize}
+          \item \textbf{DataSynthesis}, calculating the distortion for two data sets without applying other scaling ratios.
+          \item \textbf{2\_2\_affine}, calculating the distortion for two data sets with scaling ratios for parameter of interest and agent of changes.
+          \item \textbf{dilation}, calculating the distortion for two data sets with scaling ratios for parameter of interest and agent of changes.
+          \item \textbf{beta\_strain}, calculating the distortion for two data sets with scaling ratio for parameter of interest.
+          \item \textbf{omega\_strain}, calculating the distortion for two data sets with scaling ratios for agent of changes.
+          \item \textbf{identity}, calculating the distortion for two data sets with scaling ratios of 1.
```
[the remark this answers is not in the document]
Comment (yoshiurr-INL) [reply doc/user_manual/postprocessor.tex:1674]: Made new section in raven theory manual to aid the comprehension of DSS. Scaling type are noted in that specific section.

File: doc/theory_manual/raven_theory_manual.tex [hunk 24]
```diff
@@ -404,6 +404,7 @@
 \input{reducedOrderModeling.tex}
 \input{statisticalAnalysis.tex}
 \input{dataMining.tex}
+\input{dssPostProcessor.tex}
```
Comment (wangcj05) [inline doc/theory_manual/raven_theory_manual.tex:407]: This file is not uploaded to your remote branch.

File: doc/user_manual/metrics.tex (hunk 22 above)
[the remark this answers is not in the document]
Comment (wangcj05) [reply doc/user_manual/metrics.tex:610]: It seems 'attributes' is still used.

File: doc/user_manual/metrics.tex [hunk 25]
```diff
@@ -599,3 +598,32 @@ \subsubsection{Distance Based Metric}
...
+\end{itemize}
+
+An example of DSS  defined in RAVEN is provided below:
+\begin{lstlisting}[style=XML]
+<Simulation>
+  ...
+  <Metrics>
+    ...
+    <Metric name="example" subType="DSS">
+      <actionType>DataSynthetic</actionType>
```
Comment (wangcj05) [inline doc/user_manual/metrics.tex:623]: I think you have removed this node, right? It is not defined in the getInputSpecification.

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 26]
```diff
@@ -0,0 +1,412 @@
...
+    """
+    inputSpecification = super(PPDSS, cls).getInputSpecification()
+    # Have added the new pivotParameters for feature and target. The original has been commented out.
+    pivotParameterFeatureInput = InputData.parameterInputFactory("pivotParameterFeature", contentType=InputTypes.StringType,
+                                                                descr="""Pivot parameter for feature inputs""")
+    inputSpecification.addSub(pivotParameterFeatureInput)
+    pivotParameterTargetInput = InputData.parameterInputFactory("pivotParameterTarget", contentType=InputTypes.StringType,
+                                                                descr="""Pivot parameter for target inputs""")
+    inputSpecification.addSub(pivotParameterTargetInput)
+    scaleTypeInput = InputData.parameterInputFactory("scale", contentType=InputTypes.StringType,
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:60]: Please try to use ```InputTypes.makeEnumType``` to define a new type for scale parameter, see the following for an example:
image

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 27]
```diff
@@ -0,0 +1,412 @@
...
+    self.multiOutput           = 'raw_values' # defines aggregating of multiple outputs for HistorySet
+                                # currently allow raw_values
+    self.pivotParameterFeature = None
+    self.pivotValuesFeature    = []
+    self.pivotParameterTarget  = None
+    self.pivotValuesTarget     = []
+    self.scaleType             = None
+    # assembler objects to be requested
+    self.scaleRatioBeta        = []
+    self.scaleRatioOmega       = []
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:94]: please provide comments for these self variables

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 28]
```diff
@@ -0,0 +1,412 @@
...
+        timeParameter = pivotTarget
+      else:
+        timeParameter = pivotFeature
+      outputDict = {}
+      distanceTotal = np.zeros((yCount,zCount))
+      sigma = np.zeros((yCount,zCount))
+      for metric in self.metrics:
+        name = "{}_{}_{}".format(metric.estimator.name, targ.split("|")[-1], feat.split("|")[-1])
+      output = metric.evaluate((newfeatureData,newtargetData), multiOutput='raw_values')
+      #print(output)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:334]: remove print statement?

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 29]
```diff
@@ -0,0 +1,412 @@
...
+        outputDict['total_distance_'+nameTarg[1]+'_'+nameFeat[1]] = distanceTotal[cnt]
+        outputDict['feature_beta_'+nameTarg[1]+'_'+nameFeat[1]] = featureBeta[cnt]
+        outputDict['target_beta_'+nameTarg[1]+'_'+nameFeat[1]] = targetBeta[cnt]
+        outputDict['feature_omega_'+nameTarg[1]+'_'+nameFeat[1]] = featureOmegaNormScaled[cnt]
+        outputDict['target_omega_'+nameTarg[1]+'_'+nameFeat[1]] = targetOmegaNorm[cnt]
+        outputDict['feature_D_'+nameTarg[1]+'_'+nameFeat[1]] = featureD[cnt]
+        outputDict['target_D_'+nameTarg[1]+'_'+nameFeat[1]] = targetD[cnt]
+        outputDict['process_time_'+nameTarg[1]+'_'+nameFeat[1]] = newfeatureData[1][cnt]
+        outputDict['standard_deviation_'+nameTarg[1]+'_'+nameFeat[1]] = sigma[cnt]
+        #print(newfeatureData[1][cnt])
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:357]: remove this line

File: framework/Models/PostProcessors/Validations/PPDSS.py [hunk 30]
```diff
@@ -0,0 +1,412 @@
...
+    """
+      Function to place all of the computed data into the output object
+      @ In, finishedJob, JobHandler External or Internal instance, A JobHandler object that is in charge of running this post-processor
+      @ In, output, DataObject.DataObject, The object where we want to place our computed results
+      @ Out, None
+    """
+    evaluation = finishedJob.getEvaluation()
+    realizations = evaluation[1]
+    for rlz in realizations:
+      output.addRealization(rlz)
```
Comment (wangcj05) [inline framework/Models/PostProcessors/Validations/PPDSS.py:412]: I think you can use the collectOutput method in "PostProcessorInterface.py" class with key word "self.outputMultipleRealizations = True" in "__init__" method.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@yoshiurr-INL I have some additional comments for your consideration. In addition:
1. there is a missing .tex file for theory manual 
2. I am wondering if you can desire a analytic test case for DSS in future PR.

**Comment by wangcj05** [comment]
@yoshiurr-INL There are some compiling issue with the theory manual, could you check that?

**Review by wangcj05 — APPROVED** [review]
This PR is good to me.

**Pull request merged**